src/Microsoft.VisualStudio.Editor.Razor/DefaultProjectEngineFactoryService.cs

```diff
@@ -7,85 +7,177 @@
+        {
+            if (project == null)
             {
-                projectEngine = RazorProjectEngine.Create(razorConfiguration, fileSystem, b =>
-                {
-                    configure?.Invoke(b);
+                throw new ArgumentNullException(nameof(project));
+            }
 
-                    MvcLatest.RazorExtensions.Register(b);
-                });
+            if (fileSystem == null)
+            {
+                throw new ArgumentNullException(nameof(fileSystem));
             }
 
-            return projectEngine;
+            return CreateCore(project, fileSystem, configure);
         }
 
-        private ProjectSnapshot FindProject(string directory)
+        private RazorProjectEngine CreateCore(ProjectSnapshot project, RazorProjectFileSystem fileSystem, Action<RazorProjectEngineBuilder> configure)
+        {
+            // When we're running in the editor, the editor provides a configure delegate that will include
+            // the editor settings and tag helpers.
+            // 
+            // This service is only used in process in Visual Studio, and any other callers should provide these
+            // things also.
+            configure = configure ?? ((b) => { });
+
+            // The default configuration currently matches MVC-2.0. Beyond MVC-2.0 we added SDK support for 
+            // properly detecting project versions, so that's a good version to assume when we can't find a
+            // configuration.
+            var configuration = project?.Configuration ?? DefaultConfiguration;
+
+            // If there's no factory to handle the configuration then fall back to a very basic configuration.
+            //
+            // This will stop a crash from happening in this case (misconfigured project), but will still make
+            // it obvious to the user that something is wrong.
+            var factory = SelectFactory(configuration) ?? _defaultFactory;
+            return factory.Create(configuration, fileSystem, configure);
+        }
+
+        private IProjectEngineFactory SelectFactory(RazorConfiguration configuration, bool requireSerializable = false)
+        {
+            for (var i = 0; i < _customFactories.Length; i++)
+            {
+                var factory = _customFactories[i];
+                if (string.Equals(configuration.ConfigurationName, factory.Metadata.ConfigurationName))
+                {
+                    return requireSerializable && !factory.Metadata.SupportsSerialization ? null : factory.Value;
+                }
+            }
+
+            return null;
+        }
+
+        private ProjectSnapshot FindProjectByDirectory(string directory)
         {
             directory = NormalizeDirectoryPath(directory);
 
+            if (_projectManager == null)
+            {
+                _projectManager = _workspace.Services.GetLanguageServices(RazorLanguage.Name).GetRequiredService<ProjectSnapshotManager>();
```

Comment:
@rynowak, ya this is kinda hack. Don't like that our project engien factory depends on the project snapshot manager. Creates a circular dependency when constructing the objects. We might want to re-think this piece.

Reply:
Yeah this is what I wanted to remove the method that takes a path 😆

Reply:
If we can remove the method that takes a path, or move this logic into the caller, then we can delete this. I'm fine with the hack for now.